MAINT: Use intp output param viewable casts/methods

seberg opened the pull request:
This removes the cast-is-view flag with a more generic intp output.
Such an output could (theoretically) be nonzero, indicating e.g.
the offset of a complex number into a real one.
(We do not use this yet, though!)

The only "tricky" part is that the MinCastSafety helper used to
deal with the view-offset as well, and now we have to deal with it
explicitly when e.g. multiple fields have to be checked.

Bumps the experimental-dtype-api number, since the signatures changed.

---

Marked as draft, because it includes gh-20163 *not* because I think the design is "draft".  I had discussed this change with @mhvk previously e.g. in gh-19389  (Note that this does not yet use nonzero view-offsets or indicate them e.g. for certain structured casts).  These can be future enhancement, right now this is about getting the public API right.
(We should maybe try to get the indicating part right if we go the route that `arr.astype(dtype, copy=False)` return a view for more casts)

seberg committed:
MAINT: Use intp output param for descr resolution

This removes the cast-is-view flag with a more generic intp output.
Such an output could (theoretically) be nonzero, indicating e.g.
the offset of a complex number into a real one.
(We do not use this yet, though!)

The only "tricky" part is that the MinCastSafety helper used to
deal with the view-offset as well, and now we have to deal with it
explicitly when e.g. multiple fields have to be checked.

Bumps the experimental-dtype-api number, since the signatures changed.

mhvk reviewed (commented):
This looks nice, and I think overall it is a nice improvement on having the flag, even if it makes `GetCastSafety` a little odd. Also, from the many comments on code comments, you'll correctly infer that I'm not always 100% sure about what is happening...

seberg committed:
MAINT,DOC: Cleanups and clarfications based on Marten's comments

Co-Authored-By: Marten van Kerkwijk <mhvk@astro.utoronto.ca>

seberg committed:
MAINT: Rename `PyArray_GetCastSafety` to `PyArray_GetCastInfo`

Better captures the fact that it also returns the view-offset
information now.

seberg committed:
MAINT: Fix structured cast-is-view logic

seberg commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -3201,16 +3237,19 @@ PyArray_GetVoidToGenericCastingImpl(void)
     Py_ssize_t field_count = PyTuple_Size(from->names);
     if (field_count != PyTuple_Size(to->names)) {
         /* TODO: This should be rejected! */
         return NPY_UNSAFE_CASTING;
```
Moved it after this, it seemed easier.

seberg commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -2920,22 +2953,22 @@ nonstructured_to_structured_resolve_descriptors(
                 }
             }
+            if (PyTuple_Size(given_descrs[1]->names) != 1) {
+                /* A view cannot be acceptable, unset whatever the loop set */
```
OK, I fixed this for offsets, although I explicitly check for there only being one field, and just reject it (that is in theory unnecessary, but overlapping fields seem crazy enough that I think we shouldn't allow them to begin with: use `arr.view` instead).

seberg commented:
Finally updated this.  Mainly the last commit is intresting/juicy.

Getting the view offset right for fields and subarrays is a bit annoying and confusing (IMO).  And right now, it is basically unused (although of course we may want to use it for `astype()`).

Because this is confusing, I think it may make sense to focus a bit on the new tests.

mhvk commented on numpy/core/src/multiarray/convert_datatype.c:
```diff
@@ -3229,15 +3316,40 @@ can_cast_fields_safety(PyArray_Descr *from, PyArray_Descr *to)
+         * If there is one field, use its field offset.  After that propagate
+         * the view offset if they match and set to "invalid" if not.
+         */
+        if (i == 0) {
```
Could use a similar logic for the unstructured to structured case, to support the (rather ridiculous) case of multiple fields at the same offset...

seberg replied:
Yeah, I thought about it, and then thought that overlapping fields are ridiculous and should probably be discouraged (use array-views instead).  So had decided to not do it, but could look into it.

mhvk reviewed (commented):
Tricky to get back to this... But it all looks very good! My comments really are only requests for small clarifications.

seberg commented on numpy/core/src/multiarray/array_method.c:
```diff
@@ -48,13 +48,19 @@
  * We could allow setting the output descriptors specifically to simplify
  * this step.
+ *
+ * Note that the default version will indicate that the cast can be done
```
Yes, adopting your formulation.

seberg commented on numpy/core/src/multiarray/array_method.c:
```diff
@@ -570,6 +584,19 @@ boundarraymethod__resolve_descripors(
+
     /*
      * The casting flags should be the most generic casting level (except the
      * cast-is-view flag.  If no input is parametric, it must match exactly.
```
Well, this whole (private) method is currently only used for testing purposes.  So yeah, if we make this public API, we could decide to make it debug-only (although likely it is cheap enough that it doesn't matter).

seberg commented on numpy/core/tests/test_casting_unittests.py:
```diff
@@ -681,6 +712,68 @@ def test_void_and_structured_with_subarray(self, casting):
+            ("V4", "V3", 0),
+            ("V4", "V4", 0),
+            ("V3", "V4", None),
+            ("V4", "i4", None),
```
Added the comment:
```
            # Note that currently void-to-other cast goes via byte-strings
            # and is not a "view" based cast like the opposite direction:
```
the confusing thing (that I had also forgotten by now without thinking twice), is that something -> void uses "view" semantics, but void -> something goes via `bytes(element)` (so effectively uses string cast semantics).

seberg committed:
MAINT: Address review comments by Marten

seberg commented:
Thanks for looking into this!  Yeah, it is a bit unfortunate how confusing this can get, but I still like the design and concept :).

I think I addressed all comments.  Two thing are skipped:
* the case to correctly support self overlapping dtype fields, because it seems too crazy to be good.  (Frankly, I would consider that we should deprecate it.)
* That line is still not covered I think.  It is a bit hard to cover right now (it is not typical for a cast to fail in that step).  Would have to dig a bit more to figure out if there is a way...

mhvk commented:
Yes, I do think the overall design is a huge improvement! It just continues to be difficult to wrap one's head around it - but I think that just reflects it *is* quite difficult!

seberg commented:
I think we can consider this as ready to go into main.  It is barely used anyway for now, and will sit there for a bit and I believe we agree it is much better than the initial idea :).

mhvk commented:
Yes, agreed!!

seberg commented:
I think Marten's +1 should be good enough, and since it shouldn't affect much runtime and we are early in the release cycle now: I will just "self" merge this.
More comments or protest welcome, but having these PRs open makes it just a bit harder to focus on the next thing.

charris commented:
OK with me, I was going to merge it myself.
